src/twisted/internet/endpoints.py

```diff
@@ -643,6 +649,89 @@ def _resolvedHostConnect(self, resolvedHost, protocolFactory):
             return defer.fail()
 
 
+@implementer(IHostnameResolver)
+class _SimpleHostnameResolver(object):
+    """
+    An L{IHostnameResolver} provider that invokes a provided callable
+    to resolve hostnames.
+
+    @ivar _nameResolution: the callable L{resolveHostName} invokes to
+        resolve hostnames.
+    @type _nameResolution: A L{callable} with the same signature as
+    """
+    _log = Logger()
+
+    def __init__(self, nameResolution):
+        """
+        Create a L{_SimpleHostnameResolver} instance.
+        """
+        self._nameResolution = nameResolution
+
+
+    def resolveHostName(self, resolutionReceiver,
+                        hostName,
+                        portNumber=0,
+                        addressTypes=None,
+                        transportSemantics='TCP'):
+        """
+        Initiate a hostname resolution.
+
+        @param resolutionReceiver: an object that will receive each resolved
+            address as it arrives.
+        @type resolutionReceiver: L{IResolutionReceiver}
+
+        @param hostName: see interface
+
+        @param portNumber: see interface
+
+        @param addressTypes: Ignored in this implementation.
+
+        @param transportSemantics: Ignored in this implementation.
+
+        @return: The resolution in progress.
+        @rtype: L{IResolutionReceiver}
+        """
+        resolutionReceiver.resolutionBegan(HostResolution(hostName))
+        d = self._nameResolution(hostName, portNumber)
+        d.addCallback(self._deliver, resolutionReceiver)
+        d.addErrback(self._ebLog, hostName)
```

Comment:
It does - I took this approach because I had pretensions of testing each of these methods.  But coverage is good, so I might as well switch back to closures.